Add support for enhanced ecommerce

sperand-io opened the pull request:
cc/ @ndhoule (who did the vast majority of work here, kudos) & @f2prateek @hankim813 

got the validation tests passing by overriding the test's integration instance's request method with one whose superagent implementation had the requisite parser mixed in.

ndhoule commented on package.json:
```diff
@@ -23,13 +26,17 @@
   },
   "devDependencies": {
+    "eslint": "0.x",
+    "eslint-config-segment": "^1.0.17",
```
`npm install --save-dev @segment/eslint-config@3`

ndhoule commented on package.json:
```diff
@@ -23,13 +26,17 @@
     "unix-time": "^1.0.1"
   },
   "devDependencies": {
+    "eslint": "0.x",
```
change this to `2.x`

ndhoule commented on .eslintrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "segment/node"
```
fwiw this is now `"extends": "@segment/eslint-config/server"`

ndhoule commented on Makefile:
```diff
@@ -1,12 +1,16 @@
+	@npm install
+
+lint: node_modules
+	@$(ESLINT) $(SRC) $(TESTS)
```
```make
fmt: node_modules
	@$(ESLINT) -=-fix $(SRC) $(TESTS)
```

sperand-io replied:
done! and run!

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+  result.dt = page.fullName();
+  result.t = 'pageview';
+
+  if (typeof url !== 'string') return result;
```
Do you know why we do this? If so can you add a docstring here?

ndhoule replied:
This check doesn't seem to be useful

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+ *
+ * Required parameters:
+ *    - cid      (set in createcommonGAForm)
+ *    TODO: Is this option, or options?
```
any idea on this?

sperand-io replied:
just one at a time, as I understand it. how we handle checkout step options needs to be looked at across client and server; i'll make a jira ticket for that @hankim813 can investigate further. Think we should move forward with this regardless

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+ *    - v        (set in createcommonGAForm)
+ *    - pa       ('checkout') ?
+ *    - pr{x}{x} (.properties.products[x].z)
+ *    - t        ('pageview') TODO: right?
```
any idea on this?

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+ *    - cid      (set in createcommonGAForm)
+ *    - tid      (set in createcommonGAForm)
+ *    - v        (set in createcommonGAForm)
+ *    - pa       ('checkout') ?
```
I think I added this because I wasn't sure?

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+ * @return {Object}
+ */
+
+// FIXME: Multiple products?
```
?

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+ *    - tid      (set in createcommonGAForm)
+ *    - v        (set in createcommonGAForm)
+ *
+ * FIXME: Should this be a ni event?
```
any idea?

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+ *    - tid      (set in createCommonGAForm)
+ *    - v        (set in createCommonGAForm)
+ *
+ * FIXME: Should this be a ni event?
```
any idea?

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+  };
+
+  return extend(
+    // TODO: Filter where returns `undefined`
```
Any idea if this is a thing?

ndhoule commented on lib/universal/mapper.js:
```diff
@@ -0,0 +1,785 @@
+ * @return {Object}
+ */
+
+function compactObject(obj) {
```
On that TODO note above about filtering out any undefined properties—this would be a useful util for that

sperand-io replied:
good call, using it now!

ndhoule commented on test/universal.js:
```diff
@@ -195,4 +208,382 @@ describe('Google Analytics :: Universal', function(){
+    });
+
+    describe('mapper', function() {
+      // TODO: More aggressive tests
```
You can probably remove all these TODO notes

ndhoule commented:
I seem to remember something about enhanced ecommerce handling order completed events differently. I think I compared this to the client-side integration and found that there were some differences in that regard. Have you double-checked to be sure the logic here matches the client-side?

ndhoule commented:
Some notes inline, generally looks good! Thanks for getting this working!

sperand-io commented:
- updated so that all hits include `ea`, `ec`, `el`
- added support for enhanced ecommerce `Completed Order` call

- changed the "should get a 200 tests" to be "sent correct data" to reflect that they're ensuring the mapping is correct. (the subsequent validity tests ensure the equivalent of a 200 response). removed the "maps" tests as they're redundant vs. the `.sends(output)` checks in the "sent correct data" tests

sperand-io commented:
"LGTM" — @ndhoule IRL
